operator/e2e_testing/setup/shared_cluster.go

```diff
@@ -0,0 +1,483 @@
+	"github.com/docker/docker/api/types/image"
+	"github.com/docker/docker/client"
+	"github.com/sirupsen/logrus"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+)
+
+// SharedClusterManager manages a shared (singleton) k3d cluster for E2E tests
+type SharedClusterManager struct {
+	clientset                *kubernetes.Clientset
+	restConfig               *rest.Config
+	dynamicClient            dynamic.Interface
+	cleanup                  func()
+	logger                   *logrus.Logger
+	isSetup                  bool
+	workerNodes              []string
+	registryPort             string
+	relativeSkaffoldYAMLPath string
+}
+
+var (
+	sharedCluster *SharedClusterManager
+	once          sync.Once
+)
+
+// SharedCluster returns the singleton shared cluster manager
+func SharedCluster(logger *logrus.Logger, skaffoldYAMLPath string) *SharedClusterManager {
+	once.Do(func() {
+		sharedCluster = &SharedClusterManager{
+			logger:                   logger,
+			relativeSkaffoldYAMLPath: skaffoldYAMLPath,
+		}
+	})
+	return sharedCluster
+}
+
+// Setup initializes the shared cluster with maximum required resources
+func (scm *SharedClusterManager) Setup(ctx context.Context, testImages []string) error {
+	if scm.isSetup {
+		return nil
+	}
+
+	// Configuration for maximum cluster size needed (28 worker nodes + 3 server nodes)
+	customCfg := ClusterConfig{
+		Name:             "shared-e2e-test-cluster",
+		ServerNodes:      3,
+		WorkerNodes:      28,     // Maximum needed across all tests
+		WorkerMemory:     "150m", // 150m memory per agent node to fit one workload pod
+		Image:            "rancher/k3s:v1.33.5-k3s1",
+		HostPort:         "6560", // Use a different port to avoid conflicts
+		LoadBalancerPort: "8090:80",
+		EnableRegistry:   true,
+		RegistryPort:     "5001",
+		NodeLabels: []NodeLabel{
+			{
+				Key: "node_role.e2e.grove.nvidia.com",
+				// k3s refers to worker nodes as agent nodes
+				Value:       "agent",
+				NodeFilters: []string{"agent:*"},
+			},
+			// we currently don't want GPUs in e2e tests as validator is causing issues
+			{
+				Key:   "nvidia.com/gpu.deploy.operands",
+				Value: "false",
+				// k3s refers to worker nodes as agent nodes
+				NodeFilters: []string{"server:*", "agent:*"},
+			},
+		},
+		WorkerNodeTaints: []NodeTaint{
+			{
+				Key: "node_role.e2e.grove.nvidia.com",
+				// k3s refers to worker nodes as agent nodes
+				Value:  "agent",
+				Effect: "NoSchedule",
+			},
+		},
+	}
+
+	scm.registryPort = customCfg.RegistryPort
+
+	scm.logger.Info("🚀 Setting up shared k3d cluster for all e2e tests...")
+
+	restConfig, cleanup, err := SetupCompleteK3DCluster(ctx, customCfg, scm.relativeSkaffoldYAMLPath, scm.logger)
+	if err != nil {
+		return fmt.Errorf("failed to setup shared k3d cluster: %w", err)
+	}
+
```

Comment:
`Setup()` can have a named return value `err error`, and we can add this cleanup func once. This also ensures further modifications to this function in future PRs don't miss the `cleanup()` function as it is always invoked when necessary.

```suggestion
    defer func() {
        if err != nil {
            cleanup()
        }
    }()
```

Reply:
Sorry, github hid this. I'll do this in a follow up PR, tracking it.